### DISCUSSION TITLE
Provide shader printf support

### DISCUSSION THREAD

**Discussion by WangHoi** [DiligentGraphics/DiligentEngine#324] [Ideas]
If Diligent can provide some level of builtin shader printf support,
it will help shader debug alot.

It has been widely adopted since RealMJP's blog advertisement, and Blender already has shader printf support.

**Comment by TheMostDiligent** [comment]
For shader development, there is a hot shader reload feature which I find extremely useful - you can change the shader and see the effect immediately without restarting the program.

**Comment by WangHoi** [comment]
Diligent's PSO(shader) hot reload, is very helpful, but my usage scenario is a bit different:
I'm trying to calibrate light setup in Blender and Diligent Engine, 
using Blender's relief shaded map as the reference.

During light calibration, it will be more productive,
if we can printf some internal result of shading pixels of both engine, and compare.
(Diffuse/Specular/Indirect XXX)

It is nearly done, the render quality of Diligent Engine is great, you can have a look:
https://wanghoi.github.io/sichuan_map/sichuan_map.html

:cry: In hindsight, the only problem was spending too much time debugging shaders.

**Comment by TheMostDiligent** [comment]
I created an issue https://github.com/DiligentGraphics/DiligentCore/issues/634
This would certainly be useful
